models/mace/test_mace.py

```diff
@@ -129,35 +130,39 @@
 }[ase_filter]
 optim_cls: Callable[..., Optimizer] = {"FIRE": FIRE, "LBFGS": LBFGS}[ase_optimizer]
 
-for atoms in tqdm(atoms_list, desc="Relaxing"):
+for atoms in tqdm(deepcopy(atoms_list), desc="Relaxing"):
     mat_id = atoms.info[Key.mat_id]
     if mat_id in relax_results:
         continue
     try:
         atoms.calc = mace_calc
         if max_steps > 0:
-            atoms = filter_cls(atoms)
-            optimizer = optim_cls(atoms, logfile="/dev/null")
+            filtered_atoms = filter_cls(atoms)
+            optimizer = optim_cls(filtered_atoms, logfile="/dev/null")
 
             if record_traj:
-                coords, lattices = [], []
+                coords, lattices, energies = [], [], []
                 # attach observer functions to the optimizer
                 optimizer.attach(lambda: coords.append(atoms.get_positions()))  # noqa: B023
                 optimizer.attach(lambda: lattices.append(atoms.get_cell()))  # noqa: B023
+                optimizer.attach(lambda: energies.append(atoms.get_potential_energy()))  # noqa: B023
 
             optimizer.run(fmax=force_max, steps=max_steps)
         energy = atoms.get_potential_energy()  # relaxed energy
         # if max_steps > 0, atoms is wrapped by filter_cls, so extract with getattr
-        relaxed_struct = AseAtomsAdaptor.get_structure(getattr(atoms, "atoms", atoms))
+        relaxed_struct = AseAtomsAdaptor.get_structure(atoms)
         relax_results[mat_id] = {"structure": relaxed_struct, "energy": energy}
 
-        coords, lattices = (locals().get(key, []) for key in ("coords", "lattices"))
-        if record_traj and coords and lattices:
+        coords = locals().get("coords", [])
+        lattices = locals().get("lattices", [])
+        energies = locals().get("energies", [])
+        if record_traj and coords and lattices and energies:
             mace_traj = Trajectory(
-                species=relaxed_struct[mat_id].species,
+                species=atoms.get_chemical_symbols(),
```

Comment:
this should have been `relaxed_struct.species`

Reply:
I would rather be sure of the order being the same by taking it from atoms because previously the code was looking at the filtered atoms where there were a large number of extra coords even this original working implementation suffered from this flaw as the pmg `Trajectory` class doesn't check lengths at init.